src/coreclr/jit/ifconversion.cpp

```diff
@@ -558,10 +558,11 @@ bool OptIfConversionDsc::optIfConvert()
     }
 
     // Verify the test block ends with a condition that we can manipulate.
+    // Assuming that an AND connected to a JTRUE is always a valid chain.
     GenTree* last = m_startBlock->lastStmt()->GetRootNode();
     noway_assert(last->OperIs(GT_JTRUE));
     m_cond = last->gtGetOp1();
-    if (!m_cond->OperIsCompare())
+    if (!m_cond->OperIsCompare() && !m_cond->OperIs(GT_AND))
```

Comment:
Is this going to regress xarch after #81267? Why shouldn't these be if-converted for ARM64? (Isn't the if-conversion logic for AND as good as the one for branches after this PR?)

Reply:
Ah, I misread it.

I have a related question -- do we see ever see `JTRUE` with a condition that isn't a compare node before this PR? Should the new pass maybe create `NE(AND(x, y), 0)` instead to keep this invariant (if it exists)? I think that would for instance allow the xarch backend to handle the pattern efficiently as well, if we enabled it there.

Reply:
Maybe this would allow you to get rid of the odd new fgopt.cpp changes?

Reply:
>do we see ever see JTRUE with a condition that isn't a compare node before this PR?

No. You only see compares for JTRUE.

> Should the new pass maybe create NE(AND(x, y), 0) instead 

It would complicate lowering and codegen, having to look at an additional layer of node
Plus lowering would try to optimise away the `AND`.
This would probably mean getting rid of `AND_FLAGS` too.

Maybe a better solution would be a new node type, `CCMP` and it's opposite `NCCMP` (I don't like either names, but can change that later).

*Optimise bools would produce `CCMP(LT(A,B),EQ(C,D))`.
*When chaining that `CCMP` to another compare node, it would produce `CCMP(GT(E,F), AND(LT(A,B),EQ(C,D)))`.
*fgopt would invert a CCMP to a NCCMP, and vice versa.
*Lowering would look at a CCMP or NCCMP node, and try to set GT_SET_FLAGS on it.
*genCodeForJumpTrue() would check for CCMP and NCCMP. Condition would be flipped for NCCMP.
*genCodeForCCMP() would handle both CCMP and NCCMP. It would generate into flags or a register.

Would that fit for X86 too?

Reply:
> It would complicate lowering and codegen, having to look at an additional layer of node
> Plus lowering would try to optimise away the `AND`.
> This would probably mean getting rid of `AND_FLAGS` too.

I don't think it would complicate the backend significantly. The backend already has logic to handle `EQ/NE(AND(x, y), 0)`. We first transform these into `TEST_EQ/TEST_NE(x, y)`. Then further recognition transforms this into e.g. `bt` on xarch. Is it that much more complicated to add arm64 only support to transform `TEST_EQ/TEST_NE(x, y)` into `ANDFLAGS(x, y)` when the operands can emit its values directly into flags? (for `TEST_NE(x, y)` maybe a generalization of `GT_ANDFLAGS` would be needed)

> Maybe a better solution would be a new node type, CCMP and it's opposite NCCMP (I don't like either names, but can change that later).

What would be the semantics of these nodes? If-conversion/boolean opts run on HIR, and I'm very much against adding arch specific/low level nodes in the IR until we get to lowering. It is much preferable to keep the IR regular/canonical until lowering.

Reply:
Closing the loop, I suppose the backend work is then equivalent to adding containment support to `TEST_EQ/TEST_NE`?
But if we need the equivalent for `OR` then that wouldn't scale, so I think having `GT_ANDFLAGS`/`GT_ORFLAGS`/`GT_CCMP` is reasonable.

Reply:
By the way, after working on this support in the xarch backend I came to the conclusion that the containment-based approach is not scalable: https://github.com/dotnet/runtime/issues/74867#issuecomment-1408474603

It probably means that at some point we're going to have to bite the bullet and start representing these CPU flag dependencies more explicitly in the IR. At that point we're definitely going to need something like `GT_CCMP` to represent these operations that consume/define flags. So that might be a point for adding `GT_CCMP` here.

I think it would be possible to use the existing approach also today, when reordering the condition with the two operands is ok. I.e. for LIR like:
```
cond1 = LE ...
cond2 = GE  ...
chain = AND cond1, cond2
op1 = ...
op2 = ...
res = SELECT chain, op1, op2
```

we would identify the compare chain and, side effects permitting, would reorder this into:
```
op1 = ...
op2 = ...
cond1 = LE ...
cond2 = GE  ...
chain = AND cond1, cond2
res = SELECT chain, op1, op2
```

At that point we are ok to make the transformation of `chain, cond1, cond2` into `CCMP` nodes that would consume/set CPU flags, and then have a `SELECT_CC`  variant that consumes from flags (it would be much like `SELECT_HI` that exists on x86 after decomposition today).

Anyway, I don't think this work falls on your plate (and this is more like your original implementation, so I'm sorry about the detour of getting here).

Reply:
> By the way, after working on this support in the xarch backend I came to the conclusion that the containment-based approach is not scalable: [#74867 (comment)](https://github.com/dotnet/runtime/issues/74867#issuecomment-1408474603)

Having a single containing mechanism to handle both correct flags and not putting constants into a register does feel overloaded.

 
> we would identify the compare chain and, side effects permitting, would reorder this into:

This is "just" changing how next/prev is connecting the nodes?

> Anyway, I don't think this work falls on your plate (and this is more like your original implementation, so I'm sorry about the detour of getting here).

Just to confirm - your still happy for me to:
* In the optimizer, generate `NE(AND(x, y), 0)`
* In lowering, generate a `CCMP` node

How about GT_CMP_CHAIN for the new node ?

Reply:
> Having a single containing mechanism to handle both correct flags and not putting constants into a register does feel overloaded.
> 

Well, containment is quite general -- my mental model is simply that the node is handled as part of its parent. On ARM64 that's usually constants, but not always (e.g. `GT_PUTARG_STK` contains arbitrary sources too, I don't actually know if there are any other cases).
What tipped the scales for me here was the LSRA register lifetimes model that doesn't really work out once you start having to allocate internal registers for some of the contained nodes. This is probably an xarch only problem.

> This is "just" changing how next/prev is connecting the nodes?

Yes, typically just `BlockRange().Remove` + `BlockRange().InsertBefore` would be used to move them, see e.g. `MoveCFGCallArg` for an example.

> Just to confirm - your still happy for me to:
> 
> * In the optimizer, generate NE(AND(x, y), 0)
> * In lowering, generate a CCMP node
> 
> How about GT_CMP_CHAIN for the new node ?

Yes, I think this is the right approach. Is the node always going to emit the `ccmp` instruction? If so then I think `GT_CCMP` would be fitting, but I don't really have a strong opinion.